### PR TITLE
Port to PM4

### PR DESCRIPTION
**Pull request by ErikPDev**
Please make sure your issue complies with these guidelines:
- * [X] Use same formatting
- * [X] Changes must have been tested on PMMP.
- * [X] Unless it is a minor code modification, you must use an IDE.
- * [X] Have a detailed title.

### What does the PR change?

This pull request is a port to pm4.

### Testing Environment

* PocketMine-MP: 4.2.0
* PHP: 8.0.13
* Server OS: win

### Extra Information
I couldn't figure out how to Port onUpdate to api4.0.
Maybe a repeating task could work?

### CODE REVIEW

File: src/DaPigGuy/PiggyCrates/tiles/CrateTile.php [hunk 1]
```diff
@@ -9,51 +9,48 @@
 use DaPigGuy\PiggyCrates\PiggyCrates;
 use DaPigGuy\PiggyCrates\tasks\RouletteTask;
 use muqsit\invmenu\InvMenu;
-use pocketmine\command\ConsoleCommandSender;
+use muqsit\invmenu\type\InvMenuTypeIds;
+use pocketmine\block\tile\Chest;
+//use pocketmine\command\CommandSender;
```
Comment (DaPigGuy) [inline src/DaPigGuy/PiggyCrates/tiles/CrateTile.php:14]:
```suggestion
```

File: src/DaPigGuy/PiggyCrates/tasks/RouletteTask.php [hunk 2]
```diff
@@ -82,13 +84,13 @@ public function onRun(int $currentTick): void
...
-                    $this->player->getServer()->dispatchCommand(new ConsoleCommandSender(), str_replace("{PLAYER}", $this->player->getName(), $command));
+                    $this->player->getServer()->dispatchCommand(new ConsoleCommandSender($this->player->getServer(), Server::getInstance()->getLanguage()), str_replace("{PLAYER}", $this->player->getName(), $command));
                 }
                 if ($this->itemsLeft === 0) {
                     foreach ($this->crate->getCommands() as $command) {
-                        $this->player->getServer()->dispatchCommand(new ConsoleCommandSender(), str_replace("{PLAYER}", $this->player->getName(), $command));
+                        $this->player->getServer()->dispatchCommand(new ConsoleCommandSender($this->player->getServer(), Server::getInstance()->getLanguage()), str_replace("{PLAYER}", $this->player->getName(), $command));
                     }
-                    $this->player->removeWindow($this->menu->getInventory());
+                    //$this->player->removeCurrentWindow();
```
Comment (DaPigGuy) [inline src/DaPigGuy/PiggyCrates/tasks/RouletteTask.php:93]: Why is this commented out?

File: src/DaPigGuy/PiggyCrates/tasks/RouletteTask.php [hunk 3]
```diff
@@ -82,13 +84,13 @@ public function onRun(int $currentTick): void
                 $reward = $this->lastRewards[floor(self::INVENTORY_ROW_COUNT / 2)];
                 if ($reward->getType() === "item") $this->player->getInventory()->addItem($reward->getItem());
                 foreach ($reward->getCommands() as $command) {
-                    $this->player->getServer()->dispatchCommand(new ConsoleCommandSender(), str_replace("{PLAYER}", $this->player->getName(), $command));
+                    $this->player->getServer()->dispatchCommand(new ConsoleCommandSender($this->player->getServer(), Server::getInstance()->getLanguage()), str_replace("{PLAYER}", $this->player->getName(), $command));
                 }
                 if ($this->itemsLeft === 0) {
                     foreach ($this->crate->getCommands() as $command) {
-                        $this->player->getServer()->dispatchCommand(new ConsoleCommandSender(), str_replace("{PLAYER}", $this->player->getName(), $command));
+                        $this->player->getServer()->dispatchCommand(new ConsoleCommandSender($this->player->getServer(), Server::getInstance()->getLanguage()), str_replace("{PLAYER}", $this->player->getName(), $command));
```
Comment (DaPigGuy) [inline src/DaPigGuy/PiggyCrates/tasks/RouletteTask.php:91]: Please be consistent and stick to one method for accessing the Server instance.

File: src/DaPigGuy/PiggyCrates/tiles/CrateTile.php [hunk 4]
```diff
@@ -9,51 +9,48 @@
...
-    public $crateName;
+    public string $crateName;
     /** @var Crate|null */
-    public $crateType;
+    public ?Crate $crateType;
 
     /** @var bool */
-    public $isOpen = false;
+    public bool $isOpen = false;
     /** @var Player|null */
```
Comment (DaPigGuy) [inline src/DaPigGuy/PiggyCrates/tiles/CrateTile.php:34]:
```suggestion
```

File: src/DaPigGuy/PiggyCrates/tiles/CrateTile.php [hunk 5]
```diff
@@ -9,51 +9,48 @@
...
+use pocketmine\Server;
 use pocketmine\utils\TextFormat;
+use pocketmine\world\particle\FloatingTextParticle;
 
 class CrateTile extends Chest
 {
     /** @var string */
-    public $crateName;
+    public string $crateName;
     /** @var Crate|null */
```
Comment (DaPigGuy) [inline src/DaPigGuy/PiggyCrates/tiles/CrateTile.php:29]:
```suggestion
```

File: src/DaPigGuy/PiggyCrates/tiles/CrateTile.php [hunk 6]
```diff
@@ -9,51 +9,48 @@
...
-use pocketmine\tile\Chest;
+use pocketmine\network\mcpe\protocol\types\BlockPosition;
+use pocketmine\player\Player;
+use pocketmine\Server;
 use pocketmine\utils\TextFormat;
+use pocketmine\world\particle\FloatingTextParticle;
 
 class CrateTile extends Chest
 {
     /** @var string */
```
Comment (DaPigGuy) [inline src/DaPigGuy/PiggyCrates/tiles/CrateTile.php:27]:
```suggestion
```

**Review by DaPigGuy — CHANGES_REQUESTED** [review]

**Comment by DaPigGuy** [comment]
Upgrading to InvMenu 4.3.8 should also be considered

**Comment by ErikPDev** [comment]
I made a mistake, I clicked resolve convo instead of commit 😂

File: src/DaPigGuy/PiggyCrates/tasks/RouletteTask.php (hunk 2 above)
Comment (ErikPDev) [reply src/DaPigGuy/PiggyCrates/tasks/RouletteTask.php:93]: Not sure why. Must've left it out

File: src/DaPigGuy/PiggyCrates/commands/CrateCommand.php [hunk 7]
```diff
@@ -9,42 +9,42 @@
 use CortexPE\Commando\exception\ArgumentOrderException;
 use DaPigGuy\PiggyCrates\PiggyCrates;
 use pocketmine\command\CommandSender;
-use pocketmine\Player;
+use pocketmine\player\Player;
 
 class CrateCommand extends BaseCommand
 {
-    /** @var PiggyCrates */
```
[the remark this answers is not in the document]
Comment (ErikPDev) [reply src/DaPigGuy/PiggyCrates/commands/CrateCommand.php:16]: Yeah, they are used for getting messages for the lang

File: src/DaPigGuy/PiggyCrates/EventListener.php [hunk 8]
```diff
@@ -41,35 +39,22 @@ public function onInteract(PlayerInteractEvent $event): void
...
-                    $nbt = $tile->getSpawnCompound();
-                    $nbt->setString("CrateType", $crate->getName());
-                    /** @var CrateTile $newTile */
-                    $newTile = Tile::createTile("CrateTile", $level, $nbt);
-                    $newTile->spawnToAll();
-                    $tile->close();
+                    $newTile = new CrateTile($world, $block->getPosition());
+                    $newTile->setCrateType($crate);
+	                $tile->close();
+					$world->addTile($newTile);
```
[the remark this answers is not in the document]
Comment (ErikPDev) [reply src/DaPigGuy/PiggyCrates/EventListener.php]: Fixed in an earlier commit

File: src/DaPigGuy/PiggyCrates/tasks/RouletteTask.php (hunk 2 above)
Comment (ErikPDev) [reply src/DaPigGuy/PiggyCrates/tasks/RouletteTask.php]: Fixed in an earlier commit

File: src/DaPigGuy/PiggyCrates/tasks/RouletteTask.php (hunk 3 above)
Comment (ErikPDev) [reply src/DaPigGuy/PiggyCrates/tasks/RouletteTask.php:84]: Fixed in an earlier commit

**Comment by ErikPDev** [comment]
PHPStan seems to be using a PM3

**Review by ErikPDev — COMMENTED** [review]
Reviewed

**Review by DaPigGuy — APPROVED** [review]

**Review by Aericio — APPROVED** [review]

**Pull request merged**